src/nupic/types/Sdr.cpp

```diff
@@ -323,6 +323,18 @@ namespace sdr {
     }
 
 
+    void SparseDistributedRepresentation::intersection(std::vector<const SDR*> inputs) {
+        const auto &input0 = inputs[0]->getDense();
+        dense_.assign( input0.begin(), input0.end() );
+        for(auto i = 1u; i < inputs.size(); ++i) {
+            const auto &data = inputs[i]->getDense();
+            for(auto z = 0u; z < data.size(); ++z)
+                dense_[z] = dense_[z] && data[z];
```

Comment:
doing this on sparse is better on average, and much better for good (sparse) SDRs. 
- dense: 1 sample O(#D) = O(SDR::size)
- sparse: 1 sample O(#S1*#S2) ~ O(S^2) = O(SDR::sum()^2)

Reply:
The set intersection of the sparse data format should be `O( n log( n ))` where `n` is the number of active bits in the SDRs.  The procedure involves sorting everything followed by a single sweep through the sparse data.

Reply:
I'm not going to do this in this PR.  Its a big and complicated change, and it raises an issue which I don't want to deal with in this PR: should SDR.sparse be sorted?

Reply:
> I'm not going to do this in this PR. Its a big and complicated change, and it raises an issue which I don't want to deal with in this PR: should SDR.sparse be sorted?

Ok, let's think about both of these in a separate issue

Reply:
#382 for sorted SDR_sparse_t, and #383 for sparse intersection optimization